### PR TITLE
[solution] uk/medium-absolute

### PR DESCRIPTION
**Pull request by shalamai**
# Challenge [medium-absolute]

## Checklist

- [ ] Make sure that you are opening the PR with a single solution (one solution per PR)
- [ ] The solution must have at least several references for the readers to dive in depth

### CODE REVIEW

File: uk/medium-absolute.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+title: Absolute
+lang: uk
+level: medium
+tags: math template-literal
+---
+
+## Завдання
+
+Реалізуйте тип `Absolute`.
+Він має приймати `string`, `number` or `bigint`. Результатом повинно бути позитивне число у вигляді рядка.
```
Comment (bdvorianov) [inline uk/medium-absolute.md:12]:
```suggestion
Він має приймати `string`, `number` або `bigint`. Результатом повинно бути позитивне число у вигляді рядка.
```

File: uk/medium-absolute.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+Наприклад:
+
+```typescript
+type Test = -100;
+type Result = Absolute<Test>; // expected to be "100"
+```
+
+## Рішення
```
Comment (bdvorianov) [inline uk/medium-absolute.md:21]:
```suggestion
## Розв'язок
```

File: uk/medium-absolute.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+Я не жартую, в буквальному сенсі прибрати "-" з рядка.
+
+Почнемо з перевірки чи є знак "-" на початку рядка. Якщо так, то виводимо все що залишилося, використовуючи умовні типи.
+В інакшому випадку, у нас позитивне число, тому повертаємо його без змін:
+
+```typescript
+type Absolute<T extends number | string | bigint> = T extends `-${infer N}` ? N : T;
+```
+
+Таким чином, якщо вкажемо `T = “-50”`, це співпаде з нашим шаблоном `“-<N>”` де `N` буде `"50"`, й ми його повернемо.
```
Comment (bdvorianov) [inline uk/medium-absolute.md:33]:
```suggestion
Таким чином, якщо вкажемо `T = “-50”`, це відповідатиме нашому шаблону `“-<N>”` де `N` буде `"50"`, й ми його повернемо.
```

File: uk/medium-absolute.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+Та ми бачимо, що деякі з тестів все ще не проходять.
+Це тому, що ми не завжди повертаємо рядки.
+Наприклад, передавши позитивне число, воно не співпаде з рядком в умовному типі й повернеться число замість рядку.
```
Comment (bdvorianov) [inline uk/medium-absolute.md:37]:
```suggestion
Наприклад, якщо передати позитивне число, воно не збіжиться з рядком в умовному типі й повернеться число замість рядка.
```

File: uk/medium-absolute.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+Виправимо це, загорнувши тип `T` в рядковий тип-літерал, перед тим, як його повертати.
+
+```typescript
+type Absolute<T extends number | string | bigint> = T extends `-${infer N}` ? N : `${T}`;
+```
+
+Все ще не проходять тести...
+Ми не опрацьовуємо випадок, коли `T` це негативне число.
+Число знову не співпадає з нашою умовою, виведення не спрацьовує й ми повертаємо негативне число, тільки в рядку.
```
Comment (bdvorianov) [inline uk/medium-absolute.md:47]:
```suggestion
Число знову не збігається з нашою умовою, виведення не спрацьовує й ми повертаємо негативне число, тільки в рядку.
```

File: uk/medium-absolute.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+Щоб подолати це, загорнемо число в рядковий тип-літерал на початку:
+
+```typescript
+type Absolute<T extends number | string | bigint> = `${T}` extends `-${infer N}` ? N : `${T}`;
+```
+
+В результаті, ми отримали тип, який приймає `number`, `string`, `bigint` й конвертує їх в рядок.
```
Comment (bdvorianov) [inline uk/medium-absolute.md:54]:
```suggestion
В результаті, ми отримали тип, який приймає `number`, `string` або `bigint` і конвертує їх в рядок.
```

File: uk/medium-absolute.md [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+type Result = Absolute<Test>; // expected to be "100"
+```
+
+## Рішення
+
+Найлегший спосіб отримати модуль числа - це конвертувати число в рядок й прибрати знак "-".
+Я не жартую, в буквальному сенсі прибрати "-" з рядка.
+
+Почнемо з перевірки чи є знак "-" на початку рядка. Якщо так, то виводимо все що залишилося, використовуючи умовні типи.
+В інакшому випадку, у нас позитивне число, тому повертаємо його без змін:
```
Comment (bdvorianov) [inline uk/medium-absolute.md:27]:
```suggestion
В інакшому випадку, у нас додатне число, тому повертаємо його без змін:
```

File: uk/medium-absolute.md [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+Наприклад, передавши позитивне число, воно не співпаде з рядком в умовному типі й повернеться число замість рядку.
+
+Виправимо це, загорнувши тип `T` в рядковий тип-літерал, перед тим, як його повертати.
+
+```typescript
+type Absolute<T extends number | string | bigint> = T extends `-${infer N}` ? N : `${T}`;
+```
+
+Все ще не проходять тести...
+Ми не опрацьовуємо випадок, коли `T` це негативне число.
```
Comment (bdvorianov) [inline uk/medium-absolute.md:46]:
```suggestion
Ми не опрацьовуємо випадок, коли `T` це від'ємне число.
```

**Review by bdvorianov — CHANGES_REQUESTED** [review]
I've added some suggestions, please take a look. 😇

**Comment by shalamai** [comment]
agree with all of them, thanks 👍

**Review by ghaiklor — APPROVED** [review]

**Comment by ghaiklor** [comment]
@bdvorianov your turn, everything is ok with the PR? 🙂

**Comment by ghaiklor** [comment]
TravisCI are morons, disabling the auto-merge for now.

**Comment by ghaiklor** [comment]
Migrated to GitHub Actions

**Comment by ghaiklor** [comment]
@bdvorianov bump

**Review by bdvorianov — APPROVED** [review]

**Pull request merged**